perf-tool/src/agentOptions.cpp

```diff
@@ -72,56 +72,44 @@ void modifyMonitorEvents(const std::string& function, const std::string& command
         {
-            memset(&capa, 0, sizeof(jvmtiCapabilities));
             capa.can_generate_monitor_events = 1;
-
-            error = jvmti->RelinquishCapabilities(&capa);
-            check_jvmti_error(jvmti, error, "Unable to relinquish Monitor Events Capability.");
-            error = jvmti->SetEventNotificationMode(JVMTI_DISABLE, JVMTI_EVENT_MONITOR_CONTENDED_ENTERED, (jthread)NULL);
-            check_jvmti_error(jvmti, error, "Unable to disable MonitorContendedEntered event.");
-        }
-        else if (!command.compare("start"))
-        { 
-            if (verbose == INFO)
-                printf("Monitor Events Capability already enabled\n");
-            error = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_MONITOR_CONTENDED_ENTERED, (jthread)NULL);
-            check_jvmti_error(jvmti, error, "Unable to enable MonitorContendedEntered event notifications.");
-        } else {
-            invalidCommand(function, command);
         }
-    }
-    else
-    { /* cannot generate monitor events */
-        if (!command.compare("start"))
+        if (!capa.can_get_monitor_info)
         {
-            memset(&capa, 0, sizeof(jvmtiCapabilities));
-            capa.can_generate_monitor_events = 1;
-
-            error = jvmti->AddCapabilities(&capa);
-            check_jvmti_error(jvmti, error, "Unable to init Monitor Events capability.");
-
-            error = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_MONITOR_CONTENDED_ENTERED, (jthread)NULL);
-            check_jvmti_error(jvmti, error, "Unable to enable MonitorContendedEntered event notifications.");
-        }
-        else if (!command.compare("stop"))
-        { 
-            if (verbose == INFO)
-                printf("Monitor Events Capability already disabled.");
-            error = jvmti->SetEventNotificationMode(JVMTI_DISABLE, JVMTI_EVENT_MONITOR_CONTENDED_ENTERED, (jthread)NULL);
-            check_jvmti_error(jvmti, error, "Unable to disable MonitorContendedEntered event.");
-        } else {
-            invalidCommand(function, command);
+            capa.can_get_monitor_info = 1;
         }
+        error = jvmti->AddCapabilities(&capa);
+        check_jvmti_error(jvmti, error, "Unable to init Monitor Events capability.");
+        error = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_MONITOR_CONTENDED_ENTERED, (jthread)NULL);
+        check_jvmti_error(jvmti, error, "Unable to enable MonitorContendedEntered event notifications.");
+        error = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_MONITOR_CONTENDED_ENTER, (jthread)NULL);
+        check_jvmti_error(jvmti, error, "Unable to enable MonitorContendedEnter event notifications.");
+    }
+    else if (!command.compare("stop"))
+    {
+        error = jvmti->GetCapabilities(&capa);
+        check_jvmti_error(jvmti, error, "Unable to get current capabilties.");
+        error = jvmti->RelinquishCapabilities(&capa);
```

Comment:
We have to relinquish only capabilities related to monitors. If the user already started to collect other data (say object allocation events) we should not interfere with that

Reply:
@mpirvu - ok, Addressed the above review comment